mergify_engine/web/api/queues.py

```diff
@@ -312,5 +324,435 @@ async def repository_queues(
+            redis_cache, repo["owner"]["id"]
+        )
+
+        installation = context.Installation(
+            installation_json, sub, client, redis_cache, redis_train
+        )
+        repository_ctxt = installation.get_repository_from_github_data(repo)
+        config_file_scheme = await repository_ctxt.get_mergify_config_file()
+
+        if config_file_scheme and installation.redis_train:
+            config_file = get_mergify_config(config_file_scheme)
+
+            repo_id = repo["id"]
+            owner_id = installation.owner_id
+            queue_rules = config_file["queue_rules"]
+
+            if queue_name not in [rule.name for rule in queue_rules]:
+                return responses.JSONResponse(
+                    status_code=400,
+                    content={"error": f'The queue "{queue_name}" does not exists'},
+                )
+
+            queue_name_priority = queue_rules.get_prority_level(queue_name)
+            redis_hash = f"merge-pause~{owner_id}"
+
+            for rule in queue_rules:
+                redis_key = f"{repo_id}~{rule.name}"
+
+                if queue_name_priority is not None and (
+                    rule.config["priority"] > queue_name_priority
+                ):
+                    queue_pause_data_raw = await installation.redis_train.hget(
+                        redis_hash, redis_key
+                    )
+                    if queue_pause_data_raw:
+                        return responses.JSONResponse(
+                            status_code=400,
+                            content={
+                                "error": f'There is an existing paused queue "{rule.name}" above the requested queue priority level. You may use the "/unpause_from" endpoint to unpause "{queue_name}" queue from the above "{rule.name}" paused queue.'
+                            },
+                        )
+                    continue
+
+                if rule.name == queue_name:
+                    queue_pause_data_raw = await installation.redis_train.hget(
+                        redis_hash, redis_key
+                    )
+                    if queue_pause_data_raw is None:
+                        return responses.JSONResponse(
+                            status_code=400,
+                            content={
+                                "error": f'The queue "{queue_name}" is not currently on pause.'
+                            },
+                        )
+                    else:
+                        await installation.redis_train.hdel(redis_hash, redis_key)
+                        return responses.JSONResponse(
+                            status_code=200,
+                            content={
+                                "description": f'The queue "{queue_name}" has been unpaused'
+                            },
+                        )
+
+
+@router.get(
+    "/repos/{owner}/{repository}/queue/{queue_name}/unpause_from",  # noqa: FS003
```

Comment:
To me this is not a shortcut: this is the only sane way to make it work.

If you have queues `[A, B, C, D]` and you `pause` `B`, calling `unpause` on `C` and having only `C` unpaused is impossible from the design we decided on.

So there should be only one `unpause` (which is actually `unpause_from`) and that returns an error if you try to unpause `C` while the pause was set on `B`.

Reply:
Also, I implemented two different kind of unpause (1:1 and 1:many) following a case that @sileht mentioned to me: 

"Imagine a scenario in which you have two queues "high" and "low", the "low" queue only is set on pause because it is the "no refactor week", then, you have an incident and you need to pause the "high" queue, you solve your incident, but because it is still the "no refactor week", you need to unpause only the "high" queue, letting the "low" queue still on pause" 

To solve this scenario, it is somehow mandatory to have the ability to pause one and only one queue through a 1:1 endpoint, what do you think ?

Reply:
In your scenario you called:
1. first `pause low`
2. then `pause high`

to get back to the desired state you should call
3. `pause low`

🤷 

This also makes sure those pause calls are idempotent, which is a great thing.

Maybe `pause` is not the best word for all of that? We could keep `pause` for a future usage where you can `pause` a individual queue.

Maybe using the `freeze` word would make it "clearer" for this use case, and we can keep `pause` for something else? Or add a future parameter to freeze to allow to freeze/unfreeze a single queue.

Reply:
I think we can work around a better termination indeed, freeze seems better ! 

But in the case you described once you've done: 
1. first pause low
2. then pause high

To get back to the state where there is only "low" that is paused, using "unpause-from", and then re-using "pause low", there would be a window of time where there is no paused queue. Or maybe you imply that when you pause from a queue, you check if there is paused queues above, and if yes, you unpause them, in this case "high" ?